### PR TITLE
Scheduler should use a shared informer, and fix broken watch behavior for cached watches

### PR DESCRIPTION
**Pull request by smarterclayton**
Can be used either from a true shared informer or a local shared
informer created just for the scheduler.

Fixes a bug in the cache watcher where we were returning the "current" object from a watch event, not the historic event.  This means that we broke behavior when introducing the watch cache.  This may have API implications for filtering watch consumers - but on the other hand, it prevents clients filtering from seeing objects outside of their watch correctly, which can lead to other subtle bugs.

```release-note
The behavior of some watch calls to the server when filtering on fields was incorrect.  If watching objects with a filter, when an update was made that no longer matched the filter a DELETE event was correctly sent.  However, the object that was returned by that delete was not the (correct) version before the update, but instead, the newer version.  That meant the new object was not matched by the filter.  This was a regression from behavior between cached watches on the server side and uncached watches, and thus broke downstream API clients.
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
Reverts #46199

**Comment by ncdc** [comment]
@smarterclayton do you still need the `Switch the pod lister to a filtering implementation` commit?

**Comment by liggitt** [comment]
testcase?

**Comment by smarterclayton** [comment]
@kubernetes/sig-api-machinery-pr-reviews we broke watch DELETE events with filters when we introduced the watch cache.  This PR corrects that behavior... but it potentially has ramifications for API consumers.  APIs not backed by watch caches are correct (in the strict sense that they have always behaved this way since 1.0) - APIs backed by watch caches return the most recent update, not the old update.  This is an unfortunate state to be in - we are broken both ways, we regressed several releases ago, and no one noticed until I actually made code from a client that depended on the correct behavior of the system.  We need to decide whether we accept the break, fix the break, or provide both modes for those who need correct behavior.

**Comment by smarterclayton** [comment]
Yes, still need the filtering set for scheduler extensions that expect to get only assigned pods.

**Comment by smarterclayton** [comment]
Not sure what it does for final delete.

**Comment by ncdc** [comment]
@liggitt before this fix, we get (DELETED, pod.Phase=Succeeded). With the fix, we get (DELETED, pod.Phase=Running), which is correct, since the fieldSelector is for assigned, non-terminated pods. So with the fix we never get the final etcd watch record, but we shouldn't be getting that one in this case.

**Comment by smarterclayton** [comment]
I think @liggitt was referring to an unfiltered watch.

**Comment by liggitt** [comment]
> I think @liggitt was referring to an unfiltered watch.

exactly

**Comment by ncdc** [comment]
@smarterclayton would it make more sense to move the assigned pod lister functionality into [pod_expansion.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/clientset_generated/clientset/typed/core/v1/pod_expansion.go)?

**Comment by smarterclayton** [comment]
I think we might want to consider back porting the cache watch stuff if we can decide on the API implications.

**Comment by ncdc** [comment]
Send an email to the dev list? I doubt that would cover everyone using watches and relying on this incorrect behavior, but it's better than nothing.

**Comment by smarterclayton** [comment]
Consensus seems to be merge and backport.

Reviewers, do your worst

**Comment by k82cn** [comment]
@kubernetes/sig-scheduling-pr-reviews

File: plugin/pkg/scheduler/factory/factory.go [hunk 1]
```diff
@@ -132,23 +135,60 @@ func NewConfigFactory(
...
-		&v1.Pod{},
-		0,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    c.addPodToCache,
-			UpdateFunc: c.updatePodInCache,
-			DeleteFunc: c.deletePodFromCache,
+	c.scheduledPodsHasSynced = podInformer.Informer().HasSynced
+	// scheduled pod cache
+	podInformer.Informer().AddEventHandler(
+		cache.FilteringResourceEventHandler{
```
Comment (k82cn) [inline plugin/pkg/scheduler/factory/factory.go]: Will the informer cache all Pods in local (scheduler) and filter? or handled in apiserver?

**Comment by wojtek-t** [comment]
> Consensus seems to be merge and backport.

Yeah - I would recommend this too.

@smarterclayton - thanks a lot for debugging and fixing it. It's unfortunate that we didn't find this bug for 5 releases...

**Comment by wojtek-t** [comment]
This LGTM

File: plugin/pkg/scheduler/factory/factory.go (hunk 1 above)
Comment (smarterclayton) [inline plugin/pkg/scheduler/factory/factory.go]: That's the informer's job.  The informer in this package is filtered to non-terminal pods, and then split in half into the pod cache (assigned) and queue (unassigned).  The global informer in the controller manager (which this PR does not use, but an extension scheduler or hyper kube could use) would be global, and this filters down to that set.

**Comment by smarterclayton** [comment]
@kubernetes/kubernetes-release-managers this is something we want to backport to 1.6, 1.5, and any older release that we release a patch for.

**Comment by wojtek-t** [comment]
@mwielgus ^^

**Pull request merged**